File: src/modularization_programs/c/factorial/sample_factorial.c
/*
 * Author: Jhonatan Casale (jhc)
 *
 * Contact : jhonatan@jhonatancasale.com
 *         : casale.jhon@gmail.com
 *         : https://github.com/jhonatancasale
 *         : https://twitter.com/jhonatancasale
 *         : http://jhonatancasale.github.io/
 *
 * Create date Wed 23 Nov 04:14:07 BRST 2016
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <time.h>

int fact ( int );
int fact_recursive ( int );

int main (int argc, char **argv)
{
  srand ( time ( NULL ) );
  int j = ( rand () % 9 ) + 2;

  printf ( "%d! => %d\n", j, fact ( j ) );
  printf ( "%s\n", (fact ( j ) == fact_recursive ( j ) ) ? "Agree" : "Ups" );

  return (EXIT_SUCCESS);
}

int fact ( int n ) {
  int accum = 1;

  for ( int i = n; i > 1; i-- ) accum *= i;

  return accum;
}

int fact_recursive ( int n ) {
  return ( n == 1 ) ? 1 : n * fact_recursive ( n - 1 );
}

File: src/modularization_programs/python/factorial/sample_factorial.py
"""
  Author: Jhonatan Casale (jhc)
 
  Contact : jhonatan@jhonatancasale.com
          : casale.jhon@gmail.com
          : https://github.com/jhonatancasale
          : https://twitter.com/jhonatancasale
          : http://jhonatancasale.github.io/
 
  Create date Fri 25 Nov 20:18:52 BRST 2016
"""

import math

def fact(n):
    '''(int) -> int

    Precondition: n >= 0

    Return n!

    Examples:
    >>> fact(5)
    5
    >>> fact(10)
    3628800

    '''

    accum = 1

    for i in range( 1, n + 1):
        accum *= i

    return accum

def fact_recursive (n):
    '''(int) -> int

    Precondition: n >= 0

    Return n!

    Examples:
    >>> fact(5)
    5
    >>> fact(10)
    3628800

    '''

    return 1 if n == 1 else n * fact_recursive(n - 1)

def reduce_factorial (n):
    '''(int) -> int

    Precondition: n >= 0

    Return n!

    Examples:
    >>> fact(5)
    5
    >>> fact(10)
    3628800

    '''

    from functools import reduce
    return reduce((lambda x, y : x * y), range( 1, n + 1 ) )

j = 5

print('{}! = {}'.format( j, fact ( j ) ) )
print('{}'.format('Agree' if fact(j) == fact_recursive(j) else 'Ups' ) )
print('{}'.format('Really Agree' if fact(j) == math.factorial(j) else 'Ups' ) )
print('{}'.format('Really Really Agree' 
    if fact(j) == reduce_factorial(j) else 'Ups' ) )
